pkg/operator/controllers/checker/serviceprincipalchecker.go

```diff
@@ -0,0 +1,115 @@
+package checker
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+
+	"github.com/Azure/go-autorest/autorest/azure"
+	maoclient "github.com/openshift/machine-api-operator/pkg/generated/clientset/versioned"
+	"github.com/operator-framework/operator-sdk/pkg/status"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/api/validate"
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	aroclient "github.com/Azure/ARO-RP/pkg/operator/clientset/versioned"
+	"github.com/Azure/ARO-RP/pkg/operator/controllers"
+	"github.com/Azure/ARO-RP/pkg/util/aad"
+)
+
+type ServicePrincipalChecker struct {
+	log           *logrus.Entry
+	clustercli    maoclient.Interface
+	arocli        aroclient.Interface
+	kubernetescli kubernetes.Interface
+	role          string
+}
+
+func NewServicePrincipalChecker(log *logrus.Entry, maocli maoclient.Interface, arocli aroclient.Interface, kubernetescli kubernetes.Interface, role string) *ServicePrincipalChecker {
+	return &ServicePrincipalChecker{
+		log:           log,
+		clustercli:    maocli,
+		arocli:        arocli,
+		kubernetescli: kubernetescli,
+		role:          role,
+	}
+}
+
+func (r *ServicePrincipalChecker) servicePrincipalValid(ctx context.Context) error {
+	cluster, err := r.arocli.AroV1alpha1().Clusters().Get(ctx, arov1alpha1.SingletonClusterName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	resource, err := azure.ParseResourceID(cluster.Spec.ResourceID)
+	if err != nil {
+		return err
+	}
+
+	azEnv, err := azure.EnvironmentFromName(cluster.Spec.AZEnvironment)
+	if err != nil {
+		return err
+	}
+
+	azCred, err := azCredentials(ctx, r.kubernetescli)
+	if err != nil {
+		return err
+	}
+
+	token, err := aad.GetToken(ctx, r.log, string(azCred.clientID), api.SecureString(azCred.clientSecret), string(azCred.tenantID), azEnv.ActiveDirectoryEndpoint, azEnv.ResourceManagerEndpoint)
+	if err != nil {
+		return err
+	}
+
+	authorizer, err := newAuthorizer(token)
+	if err != nil {
+		return err
+	}
+
+	masterSubnetID, workerSubnetIDs, err := getSubnetIDs(ctx, cluster.Spec.VnetID, r.clustercli)
+	if err != nil {
+		return err
+	}
+
+	validator, err := validate.NewValidator(r.log, &azEnv, masterSubnetID, workerSubnetIDs, resource.SubscriptionID, authorizer)
+	if err != nil {
+		return err
+	}
+
+	err = validator.ValidateVnetPermissions(ctx)
+	if err != nil {
+		if cloudErr, ok := err.(*api.CloudError); ok && cloudErr.Code == "" {
+			cloudErr.Code = api.CloudErrorCodeInvalidServicePrincipalPermissions
+		}
+	}
+
+	return err
+}
+
+func (r *ServicePrincipalChecker) Name() string {
+	return "ServicePrincipalChecker"
+}
+
+func (r *ServicePrincipalChecker) Check(ctx context.Context) error {
```

Comment:
The contract of a "checker" is as follows:
* It is responsible for setting its conditions
* It returns an error.  If the error is nil, it will get called again after an hour.  If the error is non-nil, it will get called again sooner.

Setting the condition to true should be done as the last step of servicePrincipalValid, i.e. iff err == nil
Setting the condition to false should be done iff aad.GetToken returns an error, and should consume that error (i.e. return nil, so that we then retry after an hour)
Any other case should pass through the error without touching the condition, so that we retry again soon.

Reply:
I dont think this was addresses fully.

Reply:
I think I this is covered now.  It only updates it's condition iff any token calls fail `cond=false` or iff everything is successful `cond=true`.